### PR TITLE
tests(ISSUE-152): add e2e tests for registration, login, logout, acco…

### LINKED ISSUE #152: Add e2e tests for user account processes
## ✏️ Description: 
Add end-to-end tests for user account processes including account creation, deletion, and logout. These tests are essential to ensure full functionality and seamless operation of the system.

## 🔧 Why We Need This:
Currently, our system lacks comprehensive end-to-end tests for critical user account processes. Implementing these tests will help identify and fix potential issues, ensuring a robust and reliable user experience.

## 🚀 Proposed Solution:
Develop and integrate end-to-end tests covering account creation, deletion, and logout functionalities. These tests should simulate various scenarios and edge cases to validate the stability and performance of these key processes.

## ✅ Issue reporting checklist

- [x] I created a descriptive title for this issue
- [x] I checked that it hasn't been reported yet
- [x] I assigned the issue or added `help-wanted` label
- [x] I added appropriate labels

### PR DESCRIPTION
**Pull request by EwaRRPoland**
…unt deletion, and functionality checks

### Scope of changes:

1. Added end-to-end tests for the following user account processes:

- Account registration,
- Login,
- Logout,
- Account deletion.

2. Ensured tests cover various scenarios and edge cases to validate stability and functionality.

### How to use it:

1. Pull the latest changes from the ``main`` branch.

2. Switch to the new branch:
```
git checkout tests/ISSUE-152/register_login_logout_delete_account_tests
```
3. Run the tests:
```
npm run test:e2e
```
4. Verify that all tests pass successfully.

## Any other comments?

These tests are essential to ensure the reliability of user account processes. 

Before submitting Pull Request, I've ensured that:

- [x] The PR is associated with the relevant Issue before its creation.
- [x] All new and existing tests passed
- [x] My commit messages follow the `kawqa-gad-playwright` project's [git commit message format](https://github.com/kat-kan/kawqa-gad-playwright/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by parira2306 — CHANGES_REQUESTED** [review]
You did a lot of great work, our project will be able to expand with a lot of good tests. I hadn't seen e2e tests before, so I was happy to review them. Thanks! :)

I passed all the tests - that's great. I left a few comments in the comments on what could possibly be improved in some files. I also have a general comment - I think that to maintain the consistency of the entire project, it would be good to move all .page.ts and data.ts files from the tests folder to the appropriate folders (src/e2e/pages) and e.g. test-data/e2e.

File: tests/e2e/UI/pages/home.page.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  }
+
+  async logout(): Promise<void> {
+    await this.logoutButton.click();
+  }
+
+  async deleteUser(): Promise<void> {
+    // Handle the dialog
+    this.page.once('dialog', async (dialog) => {
+      console.log(`Dialog message: ${dialog.message()}`);
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply tests/e2e/UI/pages/home.page.ts:19]: It's worth considering and I'll include it in the next version of the tests. **Thank you very much.**

File: tests/e2e/UI/pages/login.page.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  constructor(private page: Page) {
+    this.loginInput = this.page.getByPlaceholder('Enter User Email');
+    this.passwordInput = this.page.getByPlaceholder('Enter Password');
+    this.keepMeSignedInCheckbox = this.page.getByText('keep me sign in');
+    this.loginButton = this.page.getByRole('button', { name: 'LogIn' });
+    this.userName = this.page.getByTestId('hello');
+  }
+
+  async login(email: string, password: string): Promise<void> {
+    await this.page.goto('http://localhost:3000/login/');
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply tests/e2e/UI/pages/login.page.ts:19]: Thank you, I will make the changes.

File: tests/e2e/tests/UI/login.spec.ts [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+    { tag: ['@e2e', '@registration', '@login', '@deletion'] },
+    async ({ page }) => {
+      // Arrange
+      const { firstName, lastName, email, password } = loginData1;
+
+      // Act - Register
+      await userPage.createUser(firstName, lastName, email, password);
+
+      // Assert - Registration
+      await expect(page.getByTestId('alert-popup')).toHaveText('User created');
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply tests/e2e/tests/UI/login.spec.ts:34]: Good idea, I didn't think of it ;)

File: tests/e2e/UI/pages/user.page.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  passwordInput: Locator;
+  registerButton: Locator;
+
+  constructor(private page: Page) {
+    this.firstnameInput = this.page.getByTestId('firstname-input');
+    this.lastnameInput = this.page.getByTestId('lastname-input');
+    this.emailInput = this.page.getByTestId('email-input');
+    this.passwordInput = this.page.getByTestId('password-input');
+    this.registerButton = this.page.getByTestId('register-button');
+  }
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply tests/e2e/UI/pages/user.page.ts:16]: ok :)

[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply tests/e2e/UI/pages/user.page.ts:16]: ### I added a new change. :)
### Description of changes:
- I added the function `generateRandomValidUser` in the `testDataHelpers.ts` file to generate random user data.
- I applied the new `generateRandomValidUser` function in the `UserPage` class and updated relevant tests.

File: tests/e2e/tests/UI/login.spec.ts [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+    await page.goto('/register.html');
+  });
+  // 1
+  // Retrieves user data declared in the login.data.ts file for better data control.
+  test(
+    'should register a new user successfully, log in and deletion',
+    { tag: ['@e2e', '@registration', '@login', '@deletion'] },
+    async ({ page }) => {
+      // Arrange
+      const { firstName, lastName, email, password } = loginData1;
```
Comment (kat-kan) [inline tests/e2e/tests/UI/login.spec.ts:32]: I think more meaningful name for the variable would improve readability of code
Sth like "correctLoginData" instead of just loginData1

File: test-data/e2e/UI/login.data.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+// user3
+export const loginData3 = {
+  firstName: 'Paweł',
+  lastName: 'Wiśniewski',
+  email: 'pawel.wisniewski@example.com',
+  password: 'mySecretPass3',
+};
+
+// userNew1
+export const loginDataNew1 = {
```
Comment (kat-kan) [inline test-data/e2e/UI/login.data.ts:26]: I see that these groups of login data differ because there are also avatars. So I'd try to make use of this difference while giving more meaningful information to variables, like correctLoginDataWithAvatar
I think that most of these login data are not used across tests so maybe they are not needed?

File: tests/e2e/tests/UI/login.spec.ts [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+    },
+  );
+
+  // 5 replaces test group no. 5
+  // Currently, there is no password validation - accounts can be created with single-character passwords.
+  // These tests may fail once password validation is implemented.
+
+  invalidPasswords.forEach((password) => {
+    test(
+      `should validate password requirements during registration: ${password}`,
```
Comment (kat-kan) [inline tests/e2e/tests/UI/login.spec.ts:203]: I think we should work on description of the test
When I read "should validate password requirements during registration" I expect test to check validations that DO NOT allow to create user with incorrect password. But there is no such validation.
So given how the test is implemented, I'd describe it "should register user with password ${password}`

File: tests/e2e/tests/UI/login.spec.ts [hunk 8]
```diff
@@ -0,0 +1,429 @@
+import { HomePage } from '../../../../src/e2e/pages/home.page';
```
[the remark this answers is not in the document]
Comment (kat-kan) [reply tests/e2e/tests/UI/login.spec.ts:1]: Also, maybe it would make easier to navigate through the tests if registration and login tests were in separate files?

File: tests/e2e/tests/UI/login.spec.ts [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+
+      // Assert - Login
+      await expect(page.getByTestId('login-error')).toHaveText(
+        `Invalid username or password`,
+      );
+    },
+  );
+
+  // 8
+  test(
```
Comment (kat-kan) [inline tests/e2e/tests/UI/login.spec.ts:392]: This test unfortunately fails on my end, could You please verify?

**Review by kat-kan — CHANGES_REQUESTED** [review]
Thank You for the new adventure, I'm really happy that You started contributing to KawQA big and were not afraid to commit lots of changes <3

As it was big PR, I made multiple comments, some of them are related to our team conventions, some general "clean code" remarks. The second group should be quick to implement. 
I think the most important thing is to ensure the tests pass before merging and clean code remarks etc could be also fixed in another PR.

File: .gitignore [hunk 10]
```diff
@@ -3,4 +3,5 @@ test-results/
 playwright-report/
 playwright/.cache/
 .env
-playwright-report.json
\ No newline at end of file
+playwright-report.json
+CONTRIBUTING_MINE.md
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply .gitignore:7]: Done. I wanted to have my own file, but I moved it to my repo ;)

File: .env-template
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply .env-template:1]: I'll keep an eye on it, it wasn't intentional. Locally, I have these two files: `.env`, `.env-template`. :)

File: src/e2e/pages/login.page.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
+import { Locator, Page } from '@playwright/test';
+
+export class LoginPage {
+  loginInput: Locator;
+  passwordInput: Locator;
+  loginButton: Locator;
+  keepMeSignedInCheckbox: Locator;
+  userName: Locator;
+  // assert
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply src/e2e/pages/login.page.ts:9]: I added this comment to show where I included `parira`'s suggestions. I'll remove it. :)

File: test-data/e2e/UI/test.data.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+// testData.ts
+
+export const invalidEmails = [
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply test-data/e2e/UI/test.data.ts:3]: ok :)

File: tests/e2e/tests/UI/login.spec.ts [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+
+    await page.goto('/register.html');
+  });
+  // 1
+  // Retrieves user data declared in the login.data.ts file for better data control.
+  test(
+    'should register a new user successfully, log in and deletion',
+    { tag: ['@e2e', '@registration', '@login', '@deletion'] },
+    async ({ page }) => {
+      // Arrange
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply tests/e2e/tests/UI/login.spec.ts:31]: Thank you for your feedback. I have updated all the comments from AAA (Arrange/Act/Assert) to GWT (Given/When/Then) to align with our project conventions.  :)

File: tests/e2e/tests/UI/login.spec.ts [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+
+      // Act - Register
+      await userPage.createUser(
+        userData.firstName,
+        userData.lastName,
+        userData.email,
+        userData.password,
+      );
+
+      // Assert - Registration - added locator (PR)
```
[the remark this answers is not in the document]
Comment (EwaRRPoland) [reply tests/e2e/tests/UI/login.spec.ts:171]: Thank you for the recommendation. I have removed all instances of "added locator (PR)" from the comments. :)
